src/gui/clipboard.cpp

```diff
@@ -0,0 +1,220 @@
+// Host clipboard synchronization
+// ***************************************************************************
+
+static bool text_equals(const std::string& str_1, const std::string& str_2)
+{
+	// String compare, insensitive to end of the line encoding; needed
+	// because we convert the end of line endings to host native before
+	// copying the text to the host clipboard
+
+	size_t index_1 = 0;
+	size_t index_2 = 0;
+
+	auto get_character = [](const std::string& str, size_t& index) {
+		const char result = (str[index] == '\r') ? '\n' : str[index];
+		if ((index + 1 < str.size()) &&
+		    ((str[index] == '\r' && str[index + 1] == '\n') ||
+		     (str[index] == '\n' && str[index + 1] == '\r'))) {
+			// End of the line encoded as '\r\n' or '\n\r'
+			++index;
+		}
+
+		++index;
+		return result;
+	};
+
+	while ((index_1 < str_1.size()) && (index_2 < str_2.size())) {
+		if (get_character(str_1, index_1) != get_character(str_2, index_2)) {
+			return false;
+		}
+	}
+
+	return (index_1 == str_1.size()) && (index_2 == str_2.size());
+}
+
+static void maybe_fetch_text_from_host()
+{
+	if (control->SecureMode()) {
+		// We won't need the host content for anything more
+		clipboard.host_text_utf8.clear();
+		// Do not fetch anything from host clipboard in secure mode
+		return;
+	}
+
+	bool host_content_changed = false;
+	std::string new_host_text = {};
+
+	if (SDL_FALSE == SDL_HasClipboardText()) {
+		// Host has no text in the clipboard
+		if (!clipboard.host_text_utf8.empty()) {
+			host_content_changed = true;
+		}
+	} else {
+		// Host has a text in the clipboard
+		new_host_text = SDL_GetClipboardText();
+		if (!text_equals(new_host_text, clipboard.host_text_utf8)) {
+			host_content_changed = true;
+		}
+	}
+
+	if (host_content_changed) {
+		clipboard.clear_content();
+		clipboard.host_text_utf8 = new_host_text;
+		clipboard.text_utf8      = std::move(new_host_text);
+	}
+}
+
+static void maybe_push_text_to_host()
+{
+	if (control->SecureMode()) {
+		// We won't need the host content for anything more
+		clipboard.host_text_utf8.clear();
+		// Do not push anything to host clipboard in secure mode
+		return;
+	}
+
+	// Convert end-of-the-line encoding to host native
+#if defined(WIN32)
+	const std::string newline = "\r\n";
+#else
+	const std::string newline = "\n";
+#endif
+
+	std::string converted = {};
+	converted.reserve(clipboard.text_utf8.size());
+	for (size_t index = 0; index < clipboard.text_utf8.size(); ++index) {
+		const auto character = clipboard.text_utf8[index];
+		if (character == '\r') {
+			converted += newline;
+			if (index + 1 < clipboard.text_utf8.size() &&
+			    clipboard.text_utf8[index + 1] == '\n') {
+				++index;
+			}
+		} else if (character == '\n') {
+			converted += newline;
+			if (index + 1 < clipboard.text_utf8.size() &&
+			    clipboard.text_utf8[index + 1] == '\r') {
+				++index;
+			}
+		} else {
+			converted.push_back(character);
+		}
+	}
```

Comment:
It would be better to extract the conversion part into the function and add some unit tests. Also make the line ending a parameter of the function so you can add tests for both line ending styles.

Reply:
Done, unit test created.